Devel/commands refactor

louis77 opened the pull request:
* **What does this PR do?**

Refactor the Commands list: feature specific commands are now moved into the feature source. 
Mutt_commands.c/h provides a facility for features to register their commands on init into a new dynamic Commands array. Commands for the following features were moved:

- `USE_COMP_MBOX`
- `USE_IMAP`
- `USE_LUA`
- `USE_NOTMUCH`
- `USE_SIDEBAR`

* **What are the relevant issue numbers?**

#2641

gahr commented on mutt_commands.h:
```diff
@@ -72,6 +73,11 @@ enum MuttSetCommand
 
-extern const struct Command Commands[];
+/* command registry functions */
+#define COMMANDS_REGISTER(cmds) commands_register(cmds, sizeof(cmds) / sizeof(cmds[0]))
```
We have `mutt_array_size1 in `mutt/memory.h`

louis77 replied:
Using it now, thanks for the hint.

gahr commented on mutt_commands.c:
```diff
@@ -147,18 +129,54 @@ const struct Command Commands[] = {
+
+ARRAY_HEAD(CommandsData, struct Command);
+struct CommandsData commands = ARRAY_HEAD_INITIALIZER;
+const struct Command *Commands;
```
Why do you keep this around?

louis77 replied:
I've eliminated the redundant stuff. Let me know if it looks better now.

gahr replied:
Uhm I'm still wondering why you need the global `Commands` pointer. Can't we have APIs to access the local `commands` array?

louis77 replied:
As discussed with @gahr and @flatcap on IRC remove the global `Commands` array and provider further API to access the `commands` array.

louis77 commented:
Moved to `mutt_commands.c`:
- `mutt_command_get`
- `mutt_commands_apply`

Removed global `Commands` array, replacement:
- `mutt_commands_array`

The following functions use the above:
- `mutt_get_hook_type`
- `mutt_command_complete`
- `mutt_parse_rc_buffer`

- [ ] The latter could be optimized by calling `mutt_command_get` instead
